### ISSUE TITLE
Conditional Questions on Event Types

### ISSUE THREAD

**Issue by moilejter** [calcom/cal.com#11900]
### Is your proposal related to a problem?

We would like to have the ability to ask "follow-up questions" based on the answers to previous questions in the booking request.  For example, we could have a mandatory question that asks "how did you hear about us?" with possible answers "web", "print", "social media", "personal reference" - and If they choose "web", we'd like to ask "source web site" (a text field), or if they choose "social media", we'd like to ask "social media site" (a drop down with entries for X/tweeter, Mastodon, Facebook, other).

### Describe the solution you'd like

Add an option question type, which in addition to the existing question template, asks for the parent question identifier, and a value (or a list of values) that would make this question appear below the parent question.

### Describe alternatives you've considered

We haven't come up with an alternative we liked.

### Additional context

Did not find this already suggested in either "Issues" or "Pull Requests"

### Requirement/Document

We haven't written one up.

**Comment by hubsMIT1** [comment]
After reviewing, if this feature is needed, please assign it to me, and I will work on it.

**Comment by partik03** [comment]
can I take up this one?

**Comment by jimjoice2804** [comment]
could you give a little more detail please, thank you

**Comment by LalitDeore** [comment]
Hey, have you assigned this task to anyone? If not, then I would like to go with this task. Let me give you some details about how I can create this feature.
Basically, you want to create the feature of a follow-up question based on the previous question. we can create context for each question; if the user clicks on any of the questions, we can show them a specific list of questions, and so on.

**Comment by moilejter** [comment]
Just so I know how this works - I am assuming that we have to wait for someone senior to approve this ticket, and then they would assign it to someone to work on?  (I've gotten these notices - see above - about people who'd like to work on this, bu I figured it wouldn't be my place to ask anyone to work on this, right?)

**Comment by LalitDeore** [comment]
Ok, I am new to open source, and I thought you were someone who was managing this repository and raising this issue. That's why I asked you to work on this. Thanks for clearing.

**Comment by mkos11** [comment]
Proposed Solution:

Introduce a new question type, referred to as "Conditional Question," to the booking request form. This question type would have the following attributes:

1. **Parent Question Identifier:**
   - A field where the form creator can specify the identifier of the parent question. This identifier links the conditional question to its parent question.

2. **Trigger Values:**
   - A field or set of fields where the form creator can define the trigger values that, when selected in the parent question, activate the conditional question.

3. **Question Template:**
   - Similar to existing question types, this field allows the form creator to set the text or prompt for the conditional question.

4. **Question Type:**
   - Various question types (text field, drop-down, radio buttons, etc.) that can be selected for the conditional question.

Example Scenario:

1. **Parent Question:**
   - "How did you hear about us?"
     - Options: "web," "print," "social media," "personal reference"

2. **Conditional Question 1:**
   - Type: Conditional Question
   - Parent Question Identifier: [Identifier of "How did you hear about us?"]
   - Trigger Values: "web"
   - Question Template: "Source web site"
   - Question Type: Text field

3. **Conditional Question 2:**
   - Type: Conditional Question
   - Parent Question Identifier: [Identifier of "How did you hear about us?"]
   - Trigger Values: "social media"
   - Question Template: "Social media site"
   - Question Type: Drop-down with options for X/tweeter, Mastodon, Facebook, other

This way, when a user selects "web" in the "How did you hear about us?" question, the "Source web site" question appears dynamically. Similarly, if they choose "social media," the "Social media site" question becomes visible.

Implementation Considerations:

1. **User Interface:**
   - Ensure a user-friendly interface that allows form creators to easily set up conditional questions during the form creation process.

2. **Validation:**
   - Implement proper validation to handle cases where conditional questions may depend on multiple trigger values or complex conditions.

3. **Flexibility:**
   - Design the feature to be flexible, accommodating various question types for both parent and conditional questions.

4. **Backend Logic:**
   - Develop the backend logic to dynamically render and process conditional questions based on user responses to parent questions.

By introducing this "Conditional Question" feature, the booking request form gains enhanced flexibility, enabling more personalized and context-aware data collection.

**Comment by Adarshjain1338** [comment]
Hi, Is this issue is **approved** ? and if yes, Is it assigned to anyone ?

**Comment by ahsanfarooq210** [comment]
I would like to work on this issue. please assign it to me

**Comment by Maazchogle** [comment]
Proposed Solution:

                            AI-Driven Conditional Question Generation with Reinforcement Learning

1  Intent Recognition using NLP :
   - Train a deep learning model such as BERT or RoBERTa to recognize user intents based on their responses  
   - Use pre-trained embeddings to capture semantic meaning  
   - Fine-tune the model on labeled data user responses and corresponding intents  

2  Reinforcement Learning (RL) Agent :
   - Create an RL agent that learns to dynamically generate conditional questions  
   - The RL agent interacts with the form, observes user responses, and receives rewards based on user satisfaction  
   - The agent's policy network generates questions and the reward signal guides its learning  

3  State Representation :
   - Represent the form state including user responses and context as a vector  
   - Include information about the parent question, previous answers, and any session context  

4  Action Space :
         - Define the action space for the RL agent:
         - Generate a new question with placeholders  
         - Select an existing question from a predefined set  
         - Skip generating a question  

5  Reward Function :
        - Design a reward function that balances exploration and exploitation:
        - Positive reward for relevant questions example matching intent  
        - Negative reward for irrelevant or redundant questions  
        - Encourage concise and context-aware questions  

6  Training Process :
      - Train the RL agent using Proximal Policy Optimization or other RL algorithms  
      - Simulate user interactions to collect training data  
      - Optimize the policy network to maximize cumulative rewards  

7  Dynamic Question Generation:
      - During form filling   the RL agent generates questions based on the current state  
      - It considers the parent question, user responses, and learned context  
      - Questions are dynamically tailored to user intent  

8  User Feedback Loop :
     - Collect feedback from users:
     - Did the generated question make sense?
     - Was it relevant?
     - Did it enhance the user experience?
     - Use this feedback to update the RL agent's policy  


Benefits of RL-Based Approach :

- Adaptability :
  - The RL agent learns from user interactions adapting to specific use cases  
  - It improves over time as more data is collected  

- Context Sensitivity :
  - The agent considers the entire conversation context   not just trigger values  
  - Conditional questions become contextually relevant  

- Efficient Exploration :
  - RL balances exploration trying new questions and exploitation using learned policies  
  - It avoids repetitive or unnecessary questions  

Example :  

  Cal.com Event Registration Form :


	Parent Question: How did you discover our upcoming event?

1  User Response: Via Email Newsletter
     - Intent Recognition : The AI identifies the intent as "email channel"
     - Conditional Question Generated :
     - Question : Which section of the newsletter caught your attention?
     - Question Type : Radio buttons with options like "Featured Events", "Spotlight Speakers" or "Workshops"

2     User Response: "Through Social Media Ads"  
     -   Intent Recognition  : The AI recognizes the intent as "social media advertising"
     -   Conditional Question Generated  :
     -   Question  : "Which platform did you see the ad on?"
     -   Question Type  : Drop-down with options for platforms (e.g , Facebook, Instagram, LinkedIn)  

3     User Response: "Word of Mouth"  
     -   Intent Recognition  : The AI detects the intent as "personal recommendation"
     -   Conditional Question Generated  :
     -   Question  : "Who recommended our event?"
     -   Question Type  : Text field (for entering the referrer's name)  

4     User Response: "Website Banner"  
        -   Intent Recognition  : The AI identifies the intent as "web banner"
        -   Conditional Question Generated  :
        -   Question  : "Which page on our website featured the event?"
        -   Question Type  : Text field (for specifying the page URL)  

5     User Response: "Event Listing on Community Forum"  
       -   Intent Recognition  : The AI understands the intent as "community forum"
       -   Conditional Question Generated  :
       -   Question  : "Which forum or community platform was it?"
       -   Question Type  : Text field (for providing the platform name)  

     AI-Driven Behavior:  

   - The AI dynamically generates context-aware follow-up questions based on user intent  
   - It considers the parent question and previous answers to tailor the form  
   - If the user revises their response e.g , switches from "Email Newsletter" to "Social Media Ads" , the system adapts accordingly.



  Benefits of AI-Driven Approach:  

  - Personalization :
  - Attendees receive relevant questions based on their unique journey  
  - No generic or irrelevant queries  

  - Efficiency :
  - The AI streamlines the form avoiding redundant questions  
  - Users complete registration swiftly  

  - User Satisfaction :
  - Attendees appreciate the thoughtful interaction  
  - Conditional questions enhance engagement  


Conclusion:
Offering a smooth and customized user experience building an intelligent system that dynamically produces conditional questions by integrating reinforcement learning.

**Comment by tanyasnu09** [comment]
For this do we have to suggest through a writeup or a proper code should be posted here?

**Comment by goyalaayush** [comment]
I'm interested in working on this feature. Could you please specify where exactly it needs to be implemented?

**Comment by kxhitz** [comment]
Solution Design
1. Extend the existing question model to include attributes for conditional follow-up questions. The questions model can look like this:
{
  "id": "question_id",
  "text": "How did you hear about us?",
  "type": "multiple_choice",
  "options": ["web", "print", "social media", "personal reference"],
  "is_mandatory": true,
  "follow_up": {
    "parent_id": "question_id",
    "conditions": {
      "web": {
        "type": "text",
        "question": "Source web site"
      },
      "social_media": {
        "type": "dropdown",
        "question": "Social media site",
        "options": ["X/Twitter", "Mastodon", "Facebook", "Other"]
      }
    }
  }
}
2. Follow-up Question Logic : Implement a mechanism to evaluate the answers to parent question. For each answer, check against the defined conditions and render the appropriate follow-up question if the conditions are met.
3. Front-End Implementing : 
-> Form Rendering = Use Js to dynamically render questions. When the user answers a question, the script should check if a follow-up exists for that answer and display it accordingly.
-> Event Handling  = Attach event listeners to options, so that when an option is selected, it triggers the function to check for any follow-up questions.
4. Data Collection : Ensure that the data model can store the answers to both the primary and follow-up questions. The collected data structure must look like this:
{
  "responses": {
    "how_did_you_hear": "web",
    "source_web_site": "example.com",
    "social_media_site": null
  }
}
5. Back-end Processing : Update the back-end API to handle responses to the new question types, ensuring that the responses can be saved and processes as needed.

Benefits of This Approach:
-> Enhanced User Experience
-> Efficient Data Collection 
-> Flexibility

**Comment by PaulZeliq** [comment]
Could I be assigned to this?

**Comment by talhafayyaz123** [comment]
is it still open ?

**Comment by jimkleiber** [comment]
Any update on this?

I'd really like to have a complex conditional chain for people to figure out which services they might want from me and then be able to book those services. I imagine I might be able to do this with Typeform (paid version) and then route the answers to Cal.com, but I really would just prefer to do it within Cal.com.

Anyone working on this or something similar?

**Comment by leoatflowdigital** [comment]
Unable to subscribe to notifications, so adding a comment to add a follow—also looking for this functionality.

**Comment by Likhithakathireddy** [comment]
Hi, I’d love to work on this issue. Could you please assign it to me? This will be my first contribution.

**Comment by LokendraKumar8** [comment]
can you please share the video

**Comment by Shubham13112004** [comment]
Add an optional question type that supports parent–child relationships.
Each question can have:

A parent question ID (the question it depends on)

A trigger value or list of values that cause it to appear

This way, a follow-up question only appears when the relevant parent answer is selected.

Example:

Q1: “How did you hear about us?” → Web, Print, Social Media, Personal Reference

Q2 (child of Q1, trigger = “Web”): “Source website” (text field)

Q3 (child of Q1, trigger = “Social Media”): “Social media site” (dropdown: X/Twitter, Facebook, Mastodon, Other)

Describe alternatives you've considered

Creating multiple separate forms for each type of answer (inefficient)

Using client-side scripts to manually hide/show fields (hard to maintain)
None offered a scalable or reusable approach.

Additional context
We did not find this feature already suggested in "Issues" or "Pull Requests".
This would improve form usability and allow conditional, context-aware follow-up questions.

**Comment by k-patidar** [comment]
Proposed Solutions
Several community members suggested implementations:

mkos11's approach (most detailed):

Add "Conditional Question" type with parent identifier and trigger values
Dynamic rendering based on user responses
Flexible backend logic


kxhitz's design (most practical):

json   {
     "follow_up": {
       "conditions": {
         "web": { "type": "text", "question": "Source website" }
       }
     }
   }

Maazchogle's approach (overly complex):

AI/ML-based with BERT, reinforcement learning
Likely overkill for this use case



Key Considerations

Need for parent-child question relationships
Dynamic form rendering on frontend
Data structure to store nested responses
Backend API updates for processing conditional responses
Validation logic for complex conditions

**Comment by akanksha718** [comment]
Hi 👋  
I’m new to open source and would like to work on this issue.  
Could you please assign this to me?

Thanks!

**Comment by aheranushka20-prog** [comment]
hi! I noticed this issue is still open and there's no PR linked yet.
I'm new to open source but would like to work on this.
please let me know if I can take it up .Thanks!

**Comment by Err-rr** [comment]
I see there is a draft PR. Is this still being worked on or can I contribute/improve it?

**Comment by aheranushka20-prog** [comment]
yes you can contribute. I'm not working on that issue

**Comment by aheranushka20-prog** [comment]
Hi @Err-rr, thanks for your interest!

I'm not working on this issue anymore, so feel free to take it up. You can
go ahead and improve the draft PR or create a new one if needed.

Let me know if you need any clarification. Good luck! 🚀

**Comment by prescottmarcus76-pixel** [comment]
Thanks for the update. Since you are no longer working on this issue, I
will take it from here. I'll take a look at the current draft PR and either
improve it or start a new one if necessary.

**Comment by SahilArate** [comment]
Hi 👋
I’m new to open source and would like to work on this issue.
Could you please assign this to me?

Thanks!

**Comment by harsh0641** [comment]
Hi, I'd like to work on this issue. Planning to start with a simple version one question shows up only if a specific earlier question has a specific answer. Will expand from there if needed.

**Comment by Chandan-D-C** [comment]
Hi @harsh0641, thanks for putting together this initial implementation! I’ve been looking through #11900 and the existing attempts, and I’m interested in contributing to the remaining work.

I noticed this PR currently adds the schema and form-builder configuration for parentQuestionName and triggerValue. I’m planning to trace how booking questions are rendered and see what follow-up functionality could be added without overlapping with your work.

Are you still actively working on this PR, or would you be open to contributions/follow-up work once the base approach is finalized?

**Comment by harsh0641** [comment]
Hi @Chandan-D-C, Thanks for the interest! Yes, still actively working on it. Feel free to start planning the follow-up pieces dropdown/multi-select support, cascading validation. I'll flag here once the base approach gets maintainer feedback so we don't duplicate effort.

**Comment by Chandan-D-C** [comment]
Thanks for the update! Sounds good. I'll start looking into the follow-up
pieces, particularly dropdown/multi-select support and cascading
validation, and will wait for the base approach to receive maintainer
feedback before starting any overlapping implementation.

**Comment by 101xnikhil** [comment]
Hi! I'm new to open-source contributions and I'd like to work on this issue.
I'd like to implement the conditional question support described here.
Could you please assign this issue to me? Thanks!
